orttraining/orttraining/training_api/checkpoint.h

```diff
@@ -0,0 +1,156 @@
+};
+
+/**
+ * @brief The single entry for checkpoint utilities.
+ *
+ * A checkpoint is a directory of files:
+ * checkpoint/
+ *   param_tensors.pbseq - parameter tensor protobuf messages
+ *   optim_group0_momentum0_tensors.pbseq - optimizer momentum state tensor protobuf messages
+ *   optim_group0_momentum1_tensors.pbseq - optimizer momentum state tensor protobuf messages
+ *   optim_group0_properties.pbseq - group-wise optimizer property tensor protobuf messages
+ *   custom_properties.pbseq - custom property protobuf messages
+ */
+struct CheckpointUtils {
+ public:
+  /**
+   * @brief Save ONNX initializers as ORT checkpoint.
+   *
+   * @param model_uri ONNX model file path.
+   * @param trainable_param_names trainable parameter names.
+   * @param checkpoint_path folder where checkpoint is saved.
+   * @return Status
+   */
+  static Status SaveORTCheckpoint(const std::string& model_uri,
+                                  const std::vector<std::string>& trainable_param_names,
+                                  const PathString& checkpoint_path) {
+    return OrtSaveInternal(model_uri, trainable_param_names, checkpoint_path);
+  }
+
+  /**
+   * @brief Save training states as ORT checkpoint.
+   *
+   * @param states parameter/optimizer and other user defined training states.
+   * @param checkpoint_path folder where checkpoint is saved.
+   * @return Status
+   */
+  static Status SaveORTCheckpoint(CheckpointStates& states, const PathString& checkpoint_path) {
+    return OrtSaveInternal(states, checkpoint_path);
+  }
+
+  /**
+   * @brief Load training states from ORT checkpoint.
+   *
+   * @param checkpoint_path folder where checkpoint is stored.
+   * @param checkpoint_states parameter/optimizer and other user defined training states.
+   * @return Status
+   */
+  static Status LoadORTCheckpoint(const PathString& checkpoint_path, CheckpointStates& checkpoint_states) {
+    return OrtLoadInternal(checkpoint_path, checkpoint_states);
+  }
+
+ private:
+  CheckpointUtils() {}
+
+  static Status OrtSaveInternal(const std::string& model_uri,
+                                const std::vector<std::string>& trainable_param_names,
+                                const PathString& checkpoint_path);
+
+  static Status OrtSaveModuleStatesInternal(ModuleCheckpointStates& module_states, const PathString& parameter_folder_path);
+  static Status OrtSaveOptimizerStatesInternal(OptimizerCheckpointStates& optimizer_states, const PathString& optimizer_folder_path);
+  static Status OrtSaveInternal(CheckpointStates& states, const PathString& checkpoint_path);
+
+  static Status OrtLoadModuleStatesInternal(const PathString& parameter_folder_path, ModuleCheckpointStates& module_states);
+  static Status OrtLoadOptimizerStatesInternal(const PathString& optimizer_folder_path, OptimizerCheckpointStates& optimizer_states);
+  static Status OrtLoadInternal(const PathString& checkpoint_path, CheckpointStates& checkpoint_states);
```

Comment:
Since all these are internal and static, can we define them as free functions in an anonymous namespace in the cpp file? They don't need to be a part of the header file.

Reply:
Yeah CheckpointUtils is become not an necessary as some initial implementation was changed, we just need API exposed as   normal functions.